include/seqan3/range/view/kmer_hash.hpp

```diff
@@ -12,47 +12,645 @@
+         */
+        friend shape_iterator operator-(difference_type const skip, shape_iterator const & it) noexcept
+        //!\cond
+            requires std::RandomAccessIterator<it_t>
+        //!\endcond
+        {
+            return it - skip;
+        }
+
+        /*!\anchor shape_iterator_operator-difference
+         * \brief Return offset between this and remote iterator's position.
+         * \attention This function is only avaible if `it_t` models std::RandomAccessIterator.
+         */
+        difference_type operator-(shape_iterator const & lhs) const noexcept
+        //!\cond
+            requires std::RandomAccessIterator<it_t>
+        //!\endcond
+        {
+            return static_cast<difference_type>(text_right - lhs.text_right);
+        }
+
+        /*!\brief Return offset between remote sentinel's position and this.
+         * \attention This function is only avaible if sentinel_t and it_t model std::SizedSentinel.
+         */
+        friend difference_type operator-(sentinel_t const & lhs, shape_iterator const & rhs) noexcept
+            //!\cond
+            requires std::SizedSentinel<sentinel_t, it_t>
+            //!\endcond
+        {
+            return static_cast<difference_type>(lhs - rhs.text_right);
+        }
+
+        /*!\brief Return offset this and remote sentinel's position.
+         * \attention This function is only avaible if it_t and sentinel_t model std::SizedSentinel.
+         */
+        friend difference_type operator-(shape_iterator const & lhs, sentinel_t const & rhs) noexcept
+            //!\cond
+            requires std::SizedSentinel<it_t, sentinel_t>
+            //!\endcond
+        {
+            return static_cast<difference_type>(lhs.text_right - rhs);
+        }
+
+        /*!\brief Move the iterator to a given position and return the corresponding hash value.
+         * \attention This function is only avaible if `it_t` models std::RandomAccessIterator.
+         */
+        value_type operator[](std::make_unsigned_t<difference_type> const n)
+        //!\cond
+            requires std::RandomAccessIterator<it_t>
+        //!\endcond
+        {
+            assert(pview != nullptr);
+            text_left = std::ranges::next(std::ranges::begin(*pview), n);
+            hash_full();
+            return operator*();
+        }
+
+        //!\brief Return the hash value.
+        value_type operator*() const noexcept
+        {
+            return hash_value + to_rank(*text_right);
+        }
+
+    private:
+        //!\brief The alphabet type of the passed iterator.
+        using alphabet_t = value_type_t<it_t>;
+
+        //!\brief The alphabet size.
+        static constexpr auto const sigma{alphabet_size<alphabet_t>};
+
+        //!\brief The hash value.
+        size_t hash_value{0};
+
+        //!\brief Pointer to underlying view.
+        kmer_hash_view const * pview{nullptr};
```

Comment:
If you store the shape here, you can get rid of the pointer, right? Would be the same size and less indirection.

Reply:
Mhm, I need it for random access:
`text_left = std::ranges::next(std::ranges::begin(*pview), n);`
When I jump with the `shape_iterator` I need to set this one text iterator to the corresponding position...